packages/admin/cms-admin/src/userPermissions/user/permissions/PermissionDialog.tsx

```diff
@@ -0,0 +1,282 @@
+import { gql, useApolloClient, useQuery } from "@apollo/client";
+import { CancelButton, Field, FinalForm, FinalFormInput, FinalFormSelect, FormSection, SaveButton } from "@comet/admin";
+import { FinalFormDatePicker } from "@comet/admin-date-time";
+import { CircularProgress, Dialog, DialogActions, DialogContent, DialogTitle, InputBaseProps } from "@mui/material";
+import { styled } from "@mui/material/styles";
+import React from "react";
+import { FieldRenderProps, Form } from "react-final-form";
+import { FormattedMessage } from "react-intl";
+
+import { UserPermissionsSettings } from "../../UserPermissionsPage";
+import { camelCaseToHumanReadable } from "../../utils/camelCaseToHumanReadable";
+import {
+    GQLAvailablePermissionsQuery,
+    GQLAvailablePermissionsQueryVariables,
+    GQLCreateUserPermissionMutation,
+    GQLCreateUserPermissionMutationVariables,
+    GQLPermissionQuery,
+    GQLPermissionQueryVariables,
+    GQLUpdateUserPermissionMutation,
+    GQLUpdateUserPermissionMutationVariables,
+    GQLUserPermissionDialogFragment,
+    namedOperations,
+} from "./PermissionDialog.generated";
+
+type ConfigurationProps = InputBaseProps &
+    FieldRenderProps<JSON, HTMLInputElement | HTMLTextAreaElement> & {
+        configurationComponent: () => React.ReactNode;
+    };
+
+const Configuration = ({ configurationComponent, input, disabled }: ConfigurationProps) => {
+    const ConfigurationComponent = ({ values }: { values: JSON }) => {
+        React.useEffect(() => {
+            if (input.value !== values) {
+                input.onChange(values);
+            }
+        }, [values]);
+        return configurationComponent();
+    };
+    return (
+        <FormSection title={<FormattedMessage id="comet.userPermissions.configuration" defaultMessage="Configuration" />}>
+            <Fieldset disabled={disabled}>
+                <Form
+                    onSubmit={() => {
+                        /* */
+                    }}
+                    initialValues={input.value}
+                    render={ConfigurationComponent}
+                />
+            </Fieldset>
+        </FormSection>
+    );
+};
+
+const createMutation = gql`
+    mutation CreateUserPermission(
+        $userId: String!
+        $permission: String!
+        $configuration: JSONObject
+        $validFrom: DateTime
+        $validTo: DateTime
+        $reason: String
+        $requestedBy: String
+        $approvedBy: String
+    ) {
+        userPermissionsCreatePermission(
+            data: {
```

Comment:
it's easier to pass thru $data (also typed and everything but much less code to write)

Reply:
https://github.com/vivid-planet/comet/pull/1223/commits/3d41b7594dff6331ee7f9d4c410131e7953b5beb